### PR TITLE
Add support for rate limiting

### PR DESCRIPTION
**Pull request by spyrbri**
Adding support for rate limiting.

Allow configuration of max_retries and backoff_factor.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,6 +65,17 @@ This throws error or returns `<Ping{data='pong!'}>`
...
+### Rate Limits & Exponential Backoff
+The library will keep retrying if the request exceeds the rate limit or if there's any network related error.
+By default, the request will be retried for 20 times (approximated 15 minutes) before finally giving up.
+
+You can change the retry count from the Config initializer:
+
+ * `max_retries=` sets the maximum number of retries for failed requests, default: 20
+ * `backoff_factor=` sets the exponential backoff factor, default: 2
+
+Set it to 0 to disable it.
```
Comment (bilbof) [inline README.md]: We need more detail in this sentence, should `max_retries` or `backoff_factor` be set to zero?

File: chartmogul/retry_request.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+
+METHOD_WHITELIST = ['HEAD', 'GET', 'POST', 'PUT', 'DELETE', 'OPTIONS', 'TRACE']
+STATUS_FORCELIST = (429, 500, 502, 503, 504, 520, 524)
+
+def requests_retry_session(retries=20, backoff_factor=2, session=None,):
+    session = session or requests.Session()
+    adapter = _retry_adapter(retries, backoff_factor)
+    session.mount('http://', adapter)
```
Comment (bilbof) [inline chartmogul/retry_request.py]: Will this allow requests to be made over HTTP? If so, do you think we should remove it?

File: test/api/test_retry_request.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import unittest
+import httpretty
+import chartmogul
+from unittest.mock import patch
```
Comment (bilbof) [inline test/api/test_retry_request.py]: This import doesn't seem used.

File: test/api/test_retry_request.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                httpretty.Response(body='{}', status=500),
+                httpretty.Response(body='{}', status=200),
+            ]
+        )
+
+        # max_retries set as 4
+        # backoff_factor set as 0 to avoid waiting while testing
+        config = Config("token", "secret", None, 4, 0)
+        try:
+            DataSource.create(config, data={ "test_date": date(2015, 1, 1) }).get()
```
Comment (bilbof) [inline test/api/test_retry_request.py]: This is a mock too right? Sorry I couldn't see where we mock it :)

File: README.md [hunk 5]
```diff
@@ -65,6 +65,17 @@ This throws error or returns `<Ping{data='pong!'}>`
 You can also pass to the Config initializer:
 * `request_timeout=` sets timeout for requests (seconds), default: none (see [requests docs](http://docs.python-requests.org/en/master/user/quickstart/#timeouts) for details)
 
+### Rate Limits & Exponential Backoff
+The library will keep retrying if the request exceeds the rate limit or if there's any network related error.
+By default, the request will be retried for 20 times (approximated 15 minutes) before finally giving up.
```
Comment (bilbof) [inline README.md]: 'approximated' should be 'approximately'.

File: test/api/test_retry_request.py (hunk 3 above)
Comment (spyrbri) [reply test/api/test_retry_request.py]: Yes, it was an obsolete test. Removed.

File: chartmogul/retry_request.py (hunk 2 above)
Comment (spyrbri) [reply chartmogul/retry_request.py]: Removed.

File: README.md (hunk 1 above)
Comment (spyrbri) [reply README.md]: Updated.

File: test/api/test_retry_request.py (hunk 4 above)
Comment (spyrbri) [reply test/api/test_retry_request.py]: It is mocked by httpretty here

```
31         httpretty.register_uri(
32             httpretty.POST,
33             "https://api.chartmogul.com/v1/data_sources",
              ...
```

Comment (bilbof) [reply test/api/test_retry_request.py]: Thank you!

**Comment by spyrbri** [comment]
@bilbof Do you agree with the version changed to 1.2.0 instead of 1.1.9?

**Comment by bilbof** [comment]
@spyrbri I think it's worth it. If a user has added their own retry logic for e.g. a 500 error, this may create bugs in their code - or at least, lots of additional unnecessary requests. Even if it doesn't, I don't see a negative from this small version bump.

I guess we'll need to [create a release + release doc](https://github.com/chartmogul/chartmogul-python/releases) for this.

**Pull request merged**